### DISCUSSION TITLE
Make canister calls with delegations

### DISCUSSION THREAD

**Discussion by slide-computer** [dfinity/sdk#3182] [Feature Requests]
Having the possibility to make canister calls with delegations would make it possible to use any identity in DFX indirectly including Internet Identity.

This also seems like a safer approach than exporting an identity from a wallet as PEM file. Currently I'm working on a wallet that has a non exportable identity but I would still like to enable developers to use this identity from DFX.

I'm currently thinking of a "Create delegation JSON file" screen in the wallet where user specifies an expiration which is by default a value on the safe side like 10min.

This file could then be used e.g. like this:
```bash
dfx canister call your_canister_principal your_canister_method --delegation "./delegation.json"
```

I'm open to develop the feature (when I find the time) and make a PR.

**Comment by sea-snake** [comment]
\* Intended to post this issue with this account, so commenting here to keep track of things.

**Comment by sea-snake** [comment]
I see a PR was made to agent-rs but never completed and merged: https://github.com/dfinity/agent-rs/issues/347

If there's still interest in the feature I wouldn't mind completing that work and creating a PR for DFX.

**Reply by viviveevee** [reply]
Go for it! We won't get around to doing it anytime soon
